File: code/java/generateMatrix.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int upBound = -1,leftBound = -1, rightBound = n, downBound = n;
        int val = 1;
        int i=0,j=0;
        while(val<= n * n){
            //go right
            while(j < rightBound)
                res[i][j++] = val++;
            j--;
            upBound++;
            i++;
            //go down
            while(i < downBound)
                res[i++][j] = val++;
            i--;
            rightBound--;
            j--;            
            //go left
            while(j > leftBound)
                res[i][j--] = val++;
            j++;
            downBound--;
            i--;
            
            //go up
            while(i > upBound)
                res[i--][j] = val++;
            i++;
            leftBound++;
            j++;                
        }
        return res;
    }
}

File: code/python/generateMatrix.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        upBound = leftBound = -1
        rightBound = downBound = n
        A = [0 for i in range(n)]
        res = [A.copy() for i in range(n)]
        val = 1
        i = j = 0
        while val <= n*n:
            #go right
            while j < rightBound:
                res[i][j] = val
                j = j + 1
                val = val + 1
            j = j - 1
            i = i + 1
            upBound = upBound + 1
            
            #go down
            while i < downBound:
                res[i][j] = val
                i = i + 1
                val = val + 1
            i = i - 1
            j = j - 1
            rightBound = rightBound - 1
            
            #go left
            while j > leftBound:
                res[i][j] = val
                j = j - 1
                val = val + 1
            i = i - 1
            j = j + 1
            downBound = downBound - 1
            
            #go up
            while i > upBound:
                res[i][j] = val
                i = i - 1
                val = val + 1
            i = i + 1
            j = j + 1
            leftBound = leftBound + 1
            
            
        return res
        
